WordPress/WordPressTest/Login/WordPressDotComAuthenticatorTests.swift

```diff
@@ -0,0 +1,183 @@
+            XCTFail("Unexpected successful result")
+        } catch .loginDenied {
+            // Do nothing
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+    }
+
+    @MainActor
+    func testSignInSuccess() async throws {
+        stubTokenExchange()
+        stubGetAccountDetails()
+        stubGetSites()
+
+        // Given the app is not signed in with a WP.com account
+        try XCTAssertNil(WPAccount.lookupDefaultWordPressComAccount(in: mainContext))
+
+        // When signing in with a WP.com account
+        let authenticator = WordPressDotComAuthenticator(coreDataStack: contextManager, authenticator: fakeAuthenticator(callback: ["code": "random"]))
+        let accountID = await authenticator.signIn(from: UIViewController(), context: .default)
+        XCTAssertNotNil(accountID)
+
+        // The new WP.com acount should be set as the default account.
+        let isDefaultAccount = try mainContext.existingObject(with: XCTUnwrap(accountID)).isDefaultWordPressComAccount
+        XCTAssertTrue(isDefaultAccount)
+    }
+
+    @MainActor
+    func testSignInAnotherAccount() async throws {
+        stubTokenExchange()
+        stubGetAccountDetails()
+        stubGetSites()
+
+        // Given the app is already signed in with a WP.com account.
+        let account = AccountBuilder(mainContext).with(email: "test@example.com").with(username: "default_account").with(authToken: "token").build()
+        try mainContext.save()
+        AccountService(coreDataStack: contextManager).setDefaultWordPressComAccount(account)
+
+        // When signing in with another WP.com account.
+        let authenticator = WordPressDotComAuthenticator(coreDataStack: contextManager, authenticator: fakeAuthenticator(callback: ["code": "random"]))
+        do {
+            let _ = try await authenticator.attemptSignIn(from: UIViewController(), context: .jetpackSite(accountEmail: nil))
+            XCTFail("Unexpected successful result")
+        } catch .alreadySignedIn {
+            // Do nothing
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+    }
+
+    func testSignInWithLoadingSitesError() async throws {
+        stubTokenExchange()
+        stubGetAccountDetails()
+        stubGetSitesError()
+
+        let authenticator = WordPressDotComAuthenticator(coreDataStack: contextManager, authenticator: fakeAuthenticator(callback: ["code": "random"]))
+        do {
+            let _ = try await authenticator.attemptSignIn(from: UIViewController(), context: .default)
+            XCTFail("Unexpected successful result")
+        } catch .loadingSites {
+            // Do nothing
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+    }
+}
+
+private extension WordPressDotComAuthenticatorTests {
+    func fakeAuthenticator(callback: [String: String]) -> (URL) throws(WordPressDotComAuthenticator.AuthenticationError) -> URL {
```

Comment:
(nit) I'd name it `mock`

Reply:
I wouldn't mind using either one. (not to be too pedantic here) I believe they are two different things.

From https://martinfowler.com/bliki/TestDouble.html
> **Fake** objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an [InMemoryTestDatabase](https://martinfowler.com/bliki/InMemoryTestDatabase.html) is a good example).
> **Mocks** are pre-programmed with expectations which form a specification of the calls they are expected to receive. They can throw an exception if they receive a call they don't expect and are checked during verification to ensure they got all the calls they were expecting.